File: typescript/design_pattern/mediator/main_mediator.ts
/**
 * Mediator Design Pattern in TypeScript
 *
 * The Mediator pattern defines an object that encapsulates how a set of objects interact.
 * It promotes loose coupling by keeping objects from referring to each other explicitly,
 * and lets you vary their interaction independently.
 */

// ============================================================
// Example 1: Chat Room (Classic Mediator Example)
// ============================================================

// Mediator Interface
interface ChatRoomMediator {
    showMessage(user: User, message: string): void;
    addUser(user: User): void;
    sendPrivateMessage(from: User, to: string, message: string): void;
}

// Colleague - User class
class User {
    private name: string;
    private chatRoom: ChatRoomMediator;

    constructor(name: string, chatRoom: ChatRoomMediator) {
        this.name = name;
        this.chatRoom = chatRoom;
        this.chatRoom.addUser(this);
        console.log(`[Line 27] User: ${name} joined the chat room`);
    }

    getName(): string {
        return this.name;
    }

    send(message: string): void {
        console.log(`[Line 34] ${this.name}: Sending message to chat room`);
        this.chatRoom.showMessage(this, message);
    }

    sendPrivate(toUser: string, message: string): void {
        console.log(`[Line 39] ${this.name}: Sending private message to ${toUser}`);
        this.chatRoom.sendPrivateMessage(this, toUser, message);
    }

    receive(from: string, message: string): void {
        console.log(`[Line 44] ${this.name}: Received from ${from} - "${message}"`);
    }
}

// Concrete Mediator - ChatRoom
class ChatRoom implements ChatRoomMediator {
    private users: Map<string, User> = new Map();

    addUser(user: User): void {
        this.users.set(user.getName(), user);
        console.log(`[Line 54] ChatRoom: Registered user ${user.getName()}`);
    }

    showMessage(user: User, message: string): void {
        const sender = user.getName();
        const timestamp = new Date().toLocaleTimeString();
        console.log(`[Line 60] ChatRoom: Broadcasting message from ${sender}`);

        // Broadcast to all users except sender
        this.users.forEach((recipient, name) => {
            if (name !== sender) {
                console.log(`[Line 65]   -> Delivering to ${name}`);
                recipient.receive(sender, message);
            }
        });
    }

    sendPrivateMessage(from: User, toName: string, message: string): void {
        const recipient = this.users.get(toName);
        if (recipient) {
            console.log(`[Line 74] ChatRoom: Routing private message from ${from.getName()} to ${toName}`);
            recipient.receive(from.getName(), `[Private] ${message}`);
        } else {
            console.log(`[Line 77] ChatRoom: User ${toName} not found`);
        }
    }
}

// ============================================================
// Example 2: Air Traffic Control Tower
// ============================================================

// Mediator Interface for ATC
interface ATCMediator {
    registerFlight(flight: Flight): void;
    requestLanding(flight: Flight): boolean;
    requestTakeoff(flight: Flight): boolean;
    notifyFlights(message: string, excludeFlight?: Flight): void;
}

// Colleague - Flight class
class Flight {
    private flightNumber: string;
    private atc: ATCMediator;
    private isOnGround: boolean;

    constructor(flightNumber: string, atc: ATCMediator, onGround: boolean = false) {
        this.flightNumber = flightNumber;
        this.atc = atc;
        this.isOnGround = onGround;
        this.atc.registerFlight(this);
        console.log(`[Line 102] Flight ${flightNumber}: Registered with ATC (${onGround ? 'on ground' : 'in air'})`);
    }

    getFlightNumber(): string {
        return this.flightNumber;
    }

    getIsOnGround(): boolean {
        return this.isOnGround;
    }

    setOnGround(value: boolean): void {
        this.isOnGround = value;
    }

    requestLanding(): void {
        console.log(`[Line 118] Flight ${this.flightNumber}: Requesting landing clearance`);
        const approved = this.atc.requestLanding(this);
        if (approved) {
            console.log(`[Line 121] Flight ${this.flightNumber}: Landing approved, descending...`);
            this.isOnGround = true;
        } else {
            console.log(`[Line 124] Flight ${this.flightNumber}: Landing denied, holding pattern`);
        }
    }

    requestTakeoff(): void {
        console.log(`[Line 129] Flight ${this.flightNumber}: Requesting takeoff clearance`);
        const approved = this.atc.requestTakeoff(this);
        if (approved) {
            console.log(`[Line 132] Flight ${this.flightNumber}: Takeoff approved, ascending...`);
            this.isOnGround = false;
        } else {
            console.log(`[Line 135] Flight ${this.flightNumber}: Takeoff denied, waiting at runway`);
        }
    }

    receiveNotification(message: string): void {
        console.log(`[Line 140] Flight ${this.flightNumber}: Received ATC notification - "${message}"`);
    }
}

// Concrete Mediator - Control Tower
class ControlTower implements ATCMediator {
    private flights: Map<string, Flight> = new Map();
    private runwayAvailable: boolean = true;

    registerFlight(flight: Flight): void {
        this.flights.set(flight.getFlightNumber(), flight);
        console.log(`[Line 151] ControlTower: Flight ${flight.getFlightNumber()} now under control`);
    }

    requestLanding(flight: Flight): boolean {
        console.log(`[Line 155] ControlTower: Processing landing request from ${flight.getFlightNumber()}`);

        if (!this.runwayAvailable) {
            console.log(`[Line 158] ControlTower: Runway busy, denying landing for ${flight.getFlightNumber()}`);
            return false;
        }

        this.runwayAvailable = false;
        console.log(`[Line 163] ControlTower: Runway cleared for ${flight.getFlightNumber()} landing`);

        // Notify other flights
        this.notifyFlights(`Runway in use - ${flight.getFlightNumber()} landing`, flight);

        // Simulate runway becoming available again
        setTimeout(() => {
            this.runwayAvailable = true;
            console.log(`[Line 171] ControlTower: Runway now available`);
        }, 0);

        return true;
    }

    requestTakeoff(flight: Flight): boolean {
        console.log(`[Line 178] ControlTower: Processing takeoff request from ${flight.getFlightNumber()}`);

        if (!this.runwayAvailable) {
            console.log(`[Line 181] ControlTower: Runway busy, denying takeoff for ${flight.getFlightNumber()}`);
            return false;
        }

        if (!flight.getIsOnGround()) {
            console.log(`[Line 186] ControlTower: ${flight.getFlightNumber()} is not on ground!`);
            return false;
        }

        this.runwayAvailable = false;
        console.log(`[Line 191] ControlTower: Runway cleared for ${flight.getFlightNumber()} takeoff`);

        // Notify other flights
        this.notifyFlights(`Runway in use - ${flight.getFlightNumber()} taking off`, flight);

        // Simulate runway becoming available again
        setTimeout(() => {
            this.runwayAvailable = true;
        }, 0);

        return true;
    }

    notifyFlights(message: string, excludeFlight?: Flight): void {
        console.log(`[Line 205] ControlTower: Broadcasting - "${message}"`);
        this.flights.forEach((flight) => {
            if (flight !== excludeFlight) {
                flight.receiveNotification(message);
            }
        });
    }
}

// ============================================================
// Example 3: UI Dialog Components
// ============================================================

// Mediator Interface for Dialog
interface DialogMediator {
    notify(sender: UIComponent, event: string): void;
}

// Abstract Colleague - UI Component
abstract class UIComponent {
    protected dialog: DialogMediator;
    protected name: string;

    constructor(name: string, dialog: DialogMediator) {
        this.name = name;
        this.dialog = dialog;
    }

    getName(): string {
        return this.name;
    }
}

// Concrete Colleague - Checkbox
class Checkbox extends UIComponent {
    private checked: boolean = false;

    check(): void {
        this.checked = true;
        console.log(`[Line 240] Checkbox "${this.name}": Checked`);
        this.dialog.notify(this, "check");
    }

    uncheck(): void {
        this.checked = false;
        console.log(`[Line 246] Checkbox "${this.name}": Unchecked`);
        this.dialog.notify(this, "uncheck");
    }

    isChecked(): boolean {
        return this.checked;
    }
}

// Concrete Colleague - TextInput
class TextInput extends UIComponent {
    private value: string = "";
    private enabled: boolean = true;

    setValue(value: string): void {
        this.value = value;
        console.log(`[Line 262] TextInput "${this.name}": Value set to "${value}"`);
        this.dialog.notify(this, "input");
    }

    getValue(): string {
        return this.value;
    }

    setEnabled(enabled: boolean): void {
        this.enabled = enabled;
        console.log(`[Line 272] TextInput "${this.name}": ${enabled ? 'Enabled' : 'Disabled'}`);
    }

    isEnabled(): boolean {
        return this.enabled;
    }
}

// Concrete Colleague - Button
class Button extends UIComponent {
    private enabled: boolean = true;

    click(): void {
        if (this.enabled) {
            console.log(`[Line 286] Button "${this.name}": Clicked`);
            this.dialog.notify(this, "click");
        } else {
            console.log(`[Line 289] Button "${this.name}": Click ignored (disabled)`);
        }
    }

    setEnabled(enabled: boolean): void {
        this.enabled = enabled;
        console.log(`[Line 295] Button "${this.name}": ${enabled ? 'Enabled' : 'Disabled'}`);
    }
}

// Concrete Mediator - Registration Dialog
class RegistrationDialog implements DialogMediator {
    private termsCheckbox: Checkbox;
    private emailInput: TextInput;
    private passwordInput: TextInput;
    private submitButton: Button;

    constructor() {
        console.log("[Line 307] RegistrationDialog: Initializing components");
        this.termsCheckbox = new Checkbox("Accept Terms", this);
        this.emailInput = new TextInput("Email", this);
        this.passwordInput = new TextInput("Password", this);
        this.submitButton = new Button("Submit", this);

        // Initial state
        this.submitButton.setEnabled(false);
        console.log("[Line 315] RegistrationDialog: Setup complete");
    }

    getTermsCheckbox(): Checkbox {
        return this.termsCheckbox;
    }

    getEmailInput(): TextInput {
        return this.emailInput;
    }

    getPasswordInput(): TextInput {
        return this.passwordInput;
    }

    getSubmitButton(): Button {
        return this.submitButton;
    }

    notify(sender: UIComponent, event: string): void {
        console.log(`[Line 335] RegistrationDialog: Received "${event}" from "${sender.getName()}"`);

        if (sender === this.termsCheckbox && event === "check") {
            console.log("[Line 338] RegistrationDialog: Terms accepted, enabling inputs");
            this.emailInput.setEnabled(true);
            this.passwordInput.setEnabled(true);
            this.validateForm();
        }

        if (sender === this.termsCheckbox && event === "uncheck") {
            console.log("[Line 345] RegistrationDialog: Terms rejected, disabling form");
            this.emailInput.setEnabled(false);
            this.passwordInput.setEnabled(false);
            this.submitButton.setEnabled(false);
        }

        if (sender === this.emailInput || sender === this.passwordInput) {
            this.validateForm();
        }

        if (sender === this.submitButton && event === "click") {
            console.log("[Line 356] RegistrationDialog: Form submitted!");
            console.log(`[Line 357]   Email: ${this.emailInput.getValue()}`);
            console.log(`[Line 358]   Password: ${"*".repeat(this.passwordInput.getValue().length)}`);
        }
    }

    private validateForm(): void {
        const isValid =
            this.termsCheckbox.isChecked() &&
            this.emailInput.getValue().length > 0 &&
            this.passwordInput.getValue().length > 0;

        console.log(`[Line 368] RegistrationDialog: Form validation - ${isValid ? 'valid' : 'invalid'}`);
        this.submitButton.setEnabled(isValid);
    }
}

// Demonstration
function main(): void {
    console.log("=== Mediator Pattern Demonstration ===");

    // Demo 1: Chat Room
    console.log("\n--- Chat Room Demo ---\n");

    const chatRoom = new ChatRoom();

    const alice = new User("Alice", chatRoom);
    const bob = new User("Bob", chatRoom);
    const charlie = new User("Charlie", chatRoom);

    console.log("");
    alice.send("Hello everyone!");

    console.log("");
    bob.sendPrivate("Alice", "Hey Alice, how are you?");

    console.log("");
    charlie.send("Good morning!");

    console.log("");
    bob.sendPrivate("Dave", "Are you there?"); // User not found

    // Demo 2: Air Traffic Control
    console.log("\n\n--- Air Traffic Control Demo ---\n");

    const tower = new ControlTower();

    const flight101 = new Flight("AA101", tower, false); // In air
    const flight202 = new Flight("UA202", tower, true);  // On ground
    const flight303 = new Flight("DL303", tower, false); // In air

    console.log("");
    flight101.requestLanding();  // Should be approved

    console.log("");
    flight303.requestLanding();  // Should be denied (runway busy initially, then ok)

    console.log("");
    flight202.requestTakeoff();  // Should be approved

    // Demo 3: UI Dialog
    console.log("\n\n--- UI Dialog Demo ---\n");

    const dialog = new RegistrationDialog();

    console.log("\nTrying to submit without accepting terms:");
    dialog.getSubmitButton().click(); // Should be ignored

    console.log("\nAccepting terms:");
    dialog.getTermsCheckbox().check();

    console.log("\nFilling in form:");
    dialog.getEmailInput().setValue("user@example.com");
    dialog.getPasswordInput().setValue("secret123");

    console.log("\nSubmitting form:");
    dialog.getSubmitButton().click();

    console.log("\nUnchecking terms:");
    dialog.getTermsCheckbox().uncheck();

    console.log("\nTrying to submit after unchecking terms:");
    dialog.getSubmitButton().click();

    console.log("\n=== End of Demonstration ===");
}

main();

File: rust/design_pattern/mediator/main_mediator.rs
// Mediator Design Pattern - Chat Room Example
// Demonstrates how a mediator coordinates communication between colleagues

use std::cell::RefCell;
use std::rc::Rc;

// Mediator trait - defines the interface for communication
trait Mediator {
    fn send_message(&self, message: &str, sender_id: usize);
    fn add_user(&mut self, user: Rc<RefCell<User>>);
}

// Concrete Mediator - ChatRoom coordinates all user communication
struct ChatRoom {
    name: String,
    users: Vec<Rc<RefCell<User>>>,
}

impl ChatRoom {
    fn new(name: &str) -> Self {
        println!("[ChatRoom] Creating chat room: '{}'", name);
        ChatRoom {
            name: name.to_string(),
            users: Vec::new(),
        }
    }
}

impl Mediator for ChatRoom {
    fn add_user(&mut self, user: Rc<RefCell<User>>) {
        let user_name = user.borrow().name.clone();
        println!("[ChatRoom] User '{}' joined '{}'", user_name, self.name);
        self.users.push(user);
    }

    fn send_message(&self, message: &str, sender_id: usize) {
        // Find sender name for display
        let sender_name = self.users
            .iter()
            .find(|u| u.borrow().id == sender_id)
            .map(|u| u.borrow().name.clone())
            .unwrap_or_else(|| "Unknown".to_string());

        println!("[ChatRoom] Mediating message from '{}': \"{}\"", sender_name, message);

        // Deliver message to all other users (not the sender)
        for user in &self.users {
            let user_ref = user.borrow();
            if user_ref.id != sender_id {
                println!("[ChatRoom] -> Delivering to '{}'", user_ref.name);
                drop(user_ref); // Release borrow before calling receive
                user.borrow_mut().receive(message, &sender_name);
            }
        }
    }
}

// Colleague - User who communicates through the mediator
struct User {
    id: usize,
    name: String,
    mediator: Option<Rc<RefCell<dyn Mediator>>>,
    messages_received: Vec<String>,
}

impl User {
    fn new(id: usize, name: &str) -> Rc<RefCell<Self>> {
        println!("[User] Creating user: '{}' (id: {})", name, id);
        Rc::new(RefCell::new(User {
            id,
            name: name.to_string(),
            mediator: None,
            messages_received: Vec::new(),
        }))
    }

    fn set_mediator(&mut self, mediator: Rc<RefCell<dyn Mediator>>) {
        self.mediator = Some(mediator);
    }

    fn send(&self, message: &str) {
        println!("[User] '{}' sending: \"{}\"", self.name, message);
        if let Some(mediator) = &self.mediator {
            mediator.borrow().send_message(message, self.id);
        } else {
            println!("[User] '{}' has no chat room to send message!", self.name);
        }
    }

    fn receive(&mut self, message: &str, from: &str) {
        let formatted = format!("{}: {}", from, message);
        println!("[User] '{}' received: \"{}\"", self.name, formatted);
        self.messages_received.push(formatted);
    }

    fn show_inbox(&self) {
        println!("[User] '{}' inbox ({} messages):", self.name, self.messages_received.len());
        for (i, msg) in self.messages_received.iter().enumerate() {
            println!("       {}. {}", i + 1, msg);
        }
    }
}

// Helper function to connect user to chat room
fn join_chat_room(user: &Rc<RefCell<User>>, chat_room: &Rc<RefCell<ChatRoom>>) {
    user.borrow_mut().set_mediator(chat_room.clone());
    chat_room.borrow_mut().add_user(user.clone());
}

fn main() {
    println!("=== Mediator Pattern: Chat Room Demo ===\n");

    // Create the mediator (chat room)
    let chat_room = Rc::new(RefCell::new(ChatRoom::new("Rust Developers")));

    println!();

    // Create colleagues (users)
    let alice = User::new(1, "Alice");
    let bob = User::new(2, "Bob");
    let charlie = User::new(3, "Charlie");

    println!();

    // Register users with the mediator
    println!("--- Users Joining Chat Room ---");
    join_chat_room(&alice, &chat_room);
    join_chat_room(&bob, &chat_room);
    join_chat_room(&charlie, &chat_room);

    println!("\n--- Communication Through Mediator ---\n");

    // Users communicate through the mediator
    alice.borrow().send("Hello everyone! Anyone working on design patterns?");

    println!();

    bob.borrow().send("Hi Alice! Yes, I'm implementing the Mediator pattern.");

    println!();

    charlie.borrow().send("Great topic! The Mediator helps reduce coupling.");

    println!();

    alice.borrow().send("Exactly! Objects don't need direct references to each other.");

    println!("\n--- Final Inbox Status ---\n");

    // Show what each user received
    alice.borrow().show_inbox();
    println!();
    bob.borrow().show_inbox();
    println!();
    charlie.borrow().show_inbox();

    println!("\n=== Demo Complete ===");
}
